blob-module/ios/Blob.swift

```diff
@@ -1,86 +1,125 @@
-            return self
-        }
-        
-        var dataSlice: [BlobPart] = []
-        var charsLeft = endIdx - startIdx + 1
-        for part in blobParts {
-            if part.text().count <= charsLeft {
-                charsLeft -= part.text().count
-                dataSlice.append(part)
-            } else if charsLeft > 0 {
-                var partToAdd = ""
-                for char in part.text() {
-                    if charsLeft > 0 {
-                        charsLeft -= 1
-                        partToAdd += String(char)
-                    }
-                }
-                dataSlice.append(partToAdd)
-            } else {
-                break
-            }
-        }
+    for part in blobParts {
+      let partSize = part.size()
+      if currentPos + partSize <= relativeStart {
+        currentPos += partSize
+        continue
+      }
+      if remaining <= 0 {
+        break
+      }
+      let partStart = max(0, relativeStart - currentPos)
+      let partEnd = min(partSize, partStart + remaining)
+      let length = partEnd - partStart
+      if length <= 0 {
+        currentPos += partSize
+        continue
+      }
 
-        return Blob(blobParts: dataSlice, options: options)
+      switch part {
+        case .string(let str):
+          let utf8 = Array(str.utf8)
+          let subUtf8 = Array(utf8[partStart..<partStart+length])
+          if let subStr = String(bytes: subUtf8, encoding: .utf8) {
+            dataSlice.append(.string(subStr))
+          }
+          
+        case .data(let data):
+          let subData = data.subdata(in: partStart..<partStart+length)
+          dataSlice.append(.data(subData))
+          
+        case .typedArray(let typedArray):
+          if partStart == 0 && length == typedArray.byteLength {
+            dataSlice.append(.typedArray(typedArray))
+          } else {
+            let rawPtr = typedArray.rawPointer.advanced(by: partStart)
+            let buffer = UnsafeBufferPointer<UInt8>(start: rawPtr.assumingMemoryBound(to: UInt8.self), count: length)
+            let newData = Data(buffer: buffer)
+            dataSlice.append(.data(newData))
+          }
+        
+        case .blob(let blob):
+          let subBlob = blob.slice(start: partStart, end: partStart + length)
+          dataSlice.append(.blob(subBlob))
+      }
+        currentPos += partSize
+        remaining -= length
     }
-    
-    func text() -> String {
-        return blobParts.reduce("") { $0 + $1.text() }
+  
+    let type = normalizedContentType(contentType)
+    return Blob(blobParts: dataSlice, options: BlobOptions(type: type, endings: self.options.endings))
+  }
+  
+  func text() -> String {
+    return blobParts.reduce("") { $0 + $1.text() }
+  }
+}
+
+private func normalizedContentType(_ type: String?) -> String {
+    guard let type = type, !type.isEmpty else { return "" }
+    let ascii = type.utf8.allSatisfy { $0 >= 0x20 && $0 <= 0x7E }
```

Comment:
this feels like it would be better implemented using regex, but I guess if this implementation works...

make sure to define these special characters or at least add a comment explaining them